freqtrade/exchange/kraken.py

```diff
@@ -95,32 +105,72 @@ def stoploss(self, pair: str, amount: float, stop_price: float, order_types: Dic
             logger.info('stoploss order added for %s. '
                         'stop price: %s.', pair, stop_price)
             return order
         except ccxt.InsufficientFunds as e:
             raise InsufficientFundsError(
-                f'Insufficient funds to create {ordertype} sell order on market {pair}. '
+                f'Insufficient funds to create {ordertype} {side} order on market {pair}. '
                 f'Tried to create stoploss with amount {amount} at stoploss {stop_price}. '
                 f'Message: {e}') from e
         except ccxt.InvalidOrder as e:
             raise InvalidOrderException(
-                f'Could not create {ordertype} sell order on market {pair}. '
+                f'Could not create {ordertype} {side} order on market {pair}. '
                 f'Tried to create stoploss with amount {amount} at stoploss {stop_price}. '
                 f'Message: {e}') from e
         except ccxt.DDoSProtection as e:
             raise DDosProtection(e) from e
         except (ccxt.NetworkError, ccxt.ExchangeError) as e:
             raise TemporaryError(
-                f'Could not place sell order due to {e.__class__.__name__}. Message: {e}') from e
+                f'Could not place {side} order due to {e.__class__.__name__}. Message: {e}') from e
         except ccxt.BaseError as e:
             raise OperationalException(e) from e
+
+    def fill_leverage_brackets(self):
+        """
+            Assigns property _leverage_brackets to a dictionary of information about the leverage
+            allowed on each pair
+        """
+        leverages = {}
+
+        for pair, market in self.markets.items():
+            info = market['info']
+            leverage_buy = info['leverage_buy'] if 'leverage_buy' in info else []
+            leverage_sell = info['leverage_sell'] if 'leverage_sell' in info else []
+            if len(leverage_buy) > 0 or len(leverage_sell) > 0:
+                if leverage_buy != leverage_sell:
+                    logger.warning(
+                        f"The buy({leverage_buy}) and sell({leverage_sell}) leverage are not equal"
+                        "{pair}. Please let freqtrade know because this has never happened before"
+                    )
+                    if max(leverage_buy) < max(leverage_sell):
+                        leverages[pair] = leverage_buy
+                    else:
+                        leverages[pair] = leverage_sell
+                else:
+                    leverages[pair] = leverage_buy
+        self._leverage_brackets = leverages
+
+    def get_max_leverage(self, pair: Optional[str], nominal_value: Optional[float]) -> float:
+        """
+            Returns the maximum leverage that a pair can be traded at
+            :param pair: The base/quote currency pair being traded
+            :nominal_value: Here for super class, not needed on Kraken
+        """
+        return float(max(self._leverage_brackets[pair]))
```

Comment:
i think you'll want to use `self._leverage_brackets.get(pair, <somedefault>)` (probably `[1]` here - just to be sure.
There are routes in `fill_leverage_brackets` which don't add a pair to the dict - and being safe is sometimes better ...

Reply:
so if a pair can't trade on leverage, it doesn't end up in there

Reply:
i think at the moment, nothing will end up here as it's not called anywhere except for tests.

I'd however expect that you'll have pairs (especially on kraken) which will allow leverage, and others that don't (combined in one pairlist).
There's no reason to disallow this (binance is different, as it'll be different accounts) - but on kraken, it's all "mashed together" and leverage / no leverage is just a combobox for allowed pairs.

It'll be safer / easier to call this method and get a "1" back, instead of having to somehow figure out that the pair we're currently trying to open a long position doesn't support leverage (so basically, something external to this function needs to figure out that the pair is not "leveragable" (is that a word?)).

Reply:
> "leveragable" (is that a word?)).

If you spell it leverageable it is

Reply:
This might be relevant, kraken gives an exchange error if you pass in the value 1 to params.leverage, but not if you pass in 2,3, etc.

Reply:
that's something you'll need to capture when creating the order, as "leverage" will be passed to the user, and you can't assume that you always get "non-1" values back.

That'll be exchange-specific code anyway - as for binance, you'll have to call "set_leverage=1" to disable leverage, while for kraken, it'll have to be passed in the `create_order` call (different api workings - nothing ccxt can do here, really).